src/item.rs

```diff
@@ -723,3 +750,268 @@ impl<F> Invoke<P<ast::Ty>> for ItemTyBuilder<F>
+{
+    type Result = Self;
+
+    fn invoke(self, path: ast::Path) -> Self {
+        self.with_trait(ast::TraitRef {
+            path: path,
+            ref_id: 0
+        })
+    }
+}
+
+impl<F> Invoke<P<ast::ImplItem>> for ItemImplBuilder<F>
+    where F: Invoke<P<ast::Item>>,
+{
+    type Result = Self;
+
+    fn invoke(self, item: P<ast::ImplItem>) -> Self {
+        self.with_item(item)
+    }
+}
+
+impl<F> Invoke<P<ast::Ty>> for ItemImplBuilder<F>
+    where F: Invoke<P<ast::Item>>,
+{
+    type Result = F::Result;
+
+    fn invoke(self, ty: P<ast::Ty>) -> F::Result {
+        self.build_ty(ty)
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+pub struct ItemImplItemBuilder<F> {
+    callback: F,
+    id: ast::Ident,
+    vis: ast::Visibility,
+    attrs: Vec<ast::Attribute>,
+    span: Span,
+}
+
+impl<F> ItemImplItemBuilder<F>
+    where F: Invoke<P<ast::ImplItem>>,
+{
+    pub fn new_with_callback<T>(id: T, callback: F) -> Self
+        where F: Invoke<P<ast::ImplItem>>,
+              T: ToIdent,
+    {
+        ItemImplItemBuilder {
+            callback: callback,
+            id: id.to_ident(),
+            vis: ast::Visibility::Inherited,
+            attrs: vec![],
+            span: DUMMY_SP,
+        }
+    }
+
+    pub fn span(mut self, span: Span) -> Self {
+        self.span = span;
+        self
+    }
+
+    pub fn with_attr(mut self, attr: ast::Attribute) -> Self {
+        self.attrs.push(attr);
+        self
+    }
+
+    pub fn attr(self) -> AttrBuilder<Self> {
+        AttrBuilder::new_with_callback(self)
+    }
+
+    pub fn pub_(mut self) -> Self {
+        self.vis = ast::Visibility::Public;
+        self
+    }
+
+    pub fn const_(self) -> ConstBuilder<Self> {
+        ConstBuilder::new_with_callback(self)
+    }
+
+    pub fn method(self) -> MethodBuilder<Self> {
+        MethodBuilder::new_with_callback(self)
+    }
+
+    pub fn type_(self) -> TyBuilder<Self> {
+        TyBuilder::new_with_callback(self)
+    }
+
+    pub fn mac(self) -> MacBuilder<Self> {
+        MacBuilder::new_with_callback(self)
+    }
+
+    pub fn build_item(self, node: ast::ImplItem_) -> F::Result {
+        let item = ast::ImplItem {
+            id: ast::DUMMY_NODE_ID,
+            ident: self.id,
+            vis: self.vis,
+            attrs: self.attrs,
+            node: node,
+            span: self.span,
+        };
+        self.callback.invoke(P(item))
+    }
+}
+
+impl<F> Invoke<ast::Attribute> for ItemImplItemBuilder<F>
+    where F: Invoke<P<ast::ImplItem>>,
+{
+    type Result = Self;
+
+    fn invoke(self, attr: ast::Attribute) -> Self {
+        self.with_attr(attr)
+    }
+}
+
+impl<F> Invoke<Const> for ItemImplItemBuilder<F>
+    where F: Invoke<P<ast::ImplItem>>,
+{
+    type Result = F::Result;
+
+    fn invoke(self, const_: Const) -> F::Result {
+        let node = ast::ConstImplItem(const_.ty, const_.expr.expect("an expr is required for a const impl item"));
```

Comment:
This is awesome. My only comment and I don't even know if it's worthwhile to do anymore, is that you could use multiple builders in order to make it a compile-time error if someone's created a constant that doesn't have an expression attached to it. I've done that in other places and while neat, it does add a bunch of traits and impls, so it might not be worthwhile to do. Since adding this can probably be done backwards-compatibly, I'm just going to merge this in :)